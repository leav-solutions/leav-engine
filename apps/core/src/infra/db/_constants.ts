// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export const ARANGODB_WRITEWRITE_CONFLICT_CODE = 1200;
export const ARANGODB_HTTP_CONFLICT_CODE = 409;
