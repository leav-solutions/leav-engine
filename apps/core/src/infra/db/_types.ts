// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
// Copyright LEAV Solutions 2017
// This file is released under LGPL V3

import {AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';

// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}

export interface IExecute {
    query: string | AqlQuery;
    ctx: IQueryInfos;
    withTotalCount?: boolean;
    attempts?: number;
}

export interface IExecuteWithCount<T = IDbDocument> {
    totalCount: number;
    results: T[];
}

export const isExecuteWithCount = (res: IExecuteWithCount<unknown> | unknown[]): res is IExecuteWithCount<unknown> => {
    return typeof (res as IExecuteWithCount).results !== 'undefined';
};
