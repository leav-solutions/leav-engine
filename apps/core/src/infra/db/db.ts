// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {IConfig} from '_types/config';

interface IDeps {
    config?: IConfig;
}

const _getDbConnection = (config: IConfig): Database => {
    return new Database({
        url: config.db.url
    });
};

export default function (deps: IDeps): Database {
    const db = _getDbConnection(deps.config);

    return db.database(deps.config.db.name);
}

export const initDb = async (config: IConfig) => {
    const db = _getDbConnection(config);

    const databases = await db.listDatabases();
    const dbExists = databases.reduce((exists, d) => exists || d === config.db.name, false);

    if (!dbExists) {
        await db.createDatabase(config.db.name);
    }

    db.close();
};
