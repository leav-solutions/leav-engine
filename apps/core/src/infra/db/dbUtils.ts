// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {CollectionType} from 'arangojs/lib/cjs/collection';
import {AwilixContainer} from 'awilix';
import {accessSync, constants, readdirSync} from 'fs';
import {IPluginsRepo} from 'infra/plugins/pluginsRepo';
import * as path from 'path';
import * as winston from 'winston';
import {IAttribute} from '_types/attribute';
import {IConfig} from '_types/config';
import {ILibrary} from '_types/library';
import {IList, IPaginationParams, ISortParams} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {IKeyValue} from '_types/shared';
import {ITree} from '_types/tree';
import {IDbValueVersion, IValueVersion} from '_types/value';
import {ECacheType, ICachesService} from '../../infra/cache/cacheService';
import {collectionTypes, IDbService} from './dbService';
import runMigrationFiles from './helpers/runMigrationFiles';
import {IExecuteWithCount} from './_types';

export const MIGRATIONS_COLLECTION_NAME = 'core_db_migrations';

export type CustomFilterConditionsFunc = (
    filterKey: string,
    filterVal: string | boolean | string[],
    strictFilters: boolean
) => GeneratedAqlQuery;

export interface IFindCoreEntityParams {
    collectionName: string;
    filters?: ICoreEntityFilterOptions;
    strictFilters?: boolean;
    withCount?: boolean;
    pagination?: IPaginationParams;
    sort?: ISortParams;
    customFilterConditions?: IKeyValue<CustomFilterConditionsFunc>;
    nonStrictFields?: string[];
    ctx: IQueryInfos;
}

export interface IDbUtils {
    migrate?(depsManager: AwilixContainer): Promise<void>;
    cleanup?<T extends {}>(record: {}): T;
    convertToDoc?(obj: {}): any;
    isCollectionExists?(name: string): Promise<boolean>;
    findCoreEntity?<T extends ICoreEntity>(params: IFindCoreEntityParams): Promise<IList<T>>;
    clearDatabase?(): Promise<void>;
}

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.cache.cacheService'?: ICachesService;
    'core.utils.logger'?: winston.Winston;
    'core.infra.plugins'?: IPluginsRepo;
    config?: IConfig;
}

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.cache.cacheService': cacheService = null,
    'core.utils.logger': logger = null,
    'core.infra.plugins': pluginsRepo = null,
    config = null
}: IDeps = {}): IDbUtils {
    /**
     * Create the collections used to managed db migrations
     * This can't be in a migration file because we do have to initialize it somewhere
     *
     */
    async function _initMigrationsCollection(): Promise<void> {
        const collections = await dbService.db.listCollections();

        const colExists = collections.reduce((exists, c) => exists || c.name === MIGRATIONS_COLLECTION_NAME, false);
        if (!colExists) {
            const collection = dbService.db.collection(MIGRATIONS_COLLECTION_NAME);
            await collection.create();
        }
    }

    /**
     * Return the filter's conditions based on key and val supplied.
     *
     * @param filterKey
     * @param filterVal
     * @param bindVars
     * @param index
     * @param strictFilters
     */
    function _getFilterCondition(
        filterKey: string,
        filterVal: string | boolean | string[],
        strictFilters: boolean,
        nonStrictFields?: string[]
    ): GeneratedAqlQuery {
        const queryParts = [];

        // If value is an array (types or formats for example),
        // we call this function recursively on array and join filters with an OR
        if (Array.isArray(filterVal)) {
            if (filterVal.length) {
                const valParts = filterVal.map(val => _getFilterCondition(filterKey, val, strictFilters));
                queryParts.push(aql.join(valParts, ' OR '));
            }
        } else {
            if (filterKey === 'label') {
                // Search for label in any language
                const valParts = config.lang.available.map(l => aql`LIKE(el.label.${l}, ${filterVal}, true)`);
                queryParts.push(aql.join(valParts, ' OR '));
            } else {
                // Filter with a "like" on ID or exact value in other fields
                queryParts.push(
                    (nonStrictFields ?? []).includes(filterKey) && !strictFilters
                        ? aql`LIKE(el.${filterKey}, ${filterVal}, true)`
                        : aql`el.${filterKey} == ${filterVal}`
                );
            }
        }

        return aql.join(queryParts);
    }

    const ret = {
        /**
         * Run database migrations.
         * It takes all files present in migrations folder and run it if it's never been executed before
         *
         * @param depsManager
         */
        async migrate(depsManager: AwilixContainer): Promise<void> {
            await _initMigrationsCollection();
            const ctx: IQueryInfos = {
                userId: config.defaultUserId,
                queryId: 'run-migrations'
            };
            // Load already ran migrations
            const executedMigrations = await dbService.execute<string[]>({
                query: `
                    FOR m IN core_db_migrations
                    RETURN m.file
                `,
                ctx
            });

            const _runMigrationFiles = (files, folder, prefix = null) =>
                runMigrationFiles({
                    files,
                    executedMigrations,
                    migrationsDir: folder,
                    prefix,
                    deps: {depsManager, dbService, logger},
                    ctx
                });

            /*** Core migrations ***/
            // Load migrations files
            const migrationsDir = path.resolve(__dirname, 'migrations');
            const migrationFiles = readdirSync(migrationsDir).filter(file => file.indexOf('.map') === -1);

            await _runMigrationFiles(migrationFiles, migrationsDir);

            /*** Plugins migrations ***/
            const plugins = pluginsRepo.getRegisteredPlugins();
            for (const plugin of plugins) {
                const pluginMigrationFolder = path.resolve(plugin.path + '/infra/db/migrations');

                try {
                    accessSync(pluginMigrationFolder, constants.R_OK);
                } catch (e) {
                    continue;
                }

                const pluginMigrationFiles = readdirSync(pluginMigrationFolder).filter(
                    file => file.indexOf('.map') === -1
                );

                await _runMigrationFiles(pluginMigrationFiles, pluginMigrationFolder, plugin.infos.name);
            }

            /** Clear cache */
            for (const cacheType of Object.values(ECacheType)) {
                cacheService.getCache(cacheType).deleteAll();
            }
        },

        /**
         * Cleanup every system keys from an object coming from database.
         * _key is kept under 'id'
         *
         * @param obj
         * @return any   Cleaned up object
         */
        cleanup(obj: any): any {
            if (obj === null || typeof obj === 'undefined') {
                return null;
            }

            return Object.keys(obj).reduce((newObj: any, key) => {
                if (key === '_key') {
                    newObj.id = obj[key];
                } else if (key[0] !== '_') {
                    newObj[key] = obj[key];
                }

                return newObj;
            }, {});
        },

        /**
         * Convert an object to an object looking like a DB document
         * id is replaced by _key
         *
         * @param obj
         * @return any   DB document compatible object
         */
        convertToDoc(obj: {}): any {
            const newObj: any = {...obj};

            if (typeof newObj.id !== 'undefined') {
                newObj._key = newObj.id;
            }
            delete newObj.id;

            return newObj;
        },
        /**
         * Search core entities (libraries, attributes, trees)
         *
         * @param collectionName
         * @param filters
         * @param strictFilters
         */
        async findCoreEntity<T extends ITree | ILibrary | IAttribute>(
            params: IFindCoreEntityParams
        ): Promise<IList<T>> {
            const {
                collectionName = null,
                filters = null,
                strictFilters = false,
                withCount = false,
                pagination = null,
                sort = null,
                customFilterConditions = {},
                nonStrictFields = ['label', '_key'],
                ctx = {}
            } = params;

            const collec = dbService.db.collection(collectionName);
            const queryParts = [aql`FOR el IN ${collec}`];

            if (filters !== null) {
                const dbFilters = ret.convertToDoc(filters);
                const filtersKeys = Object.keys(dbFilters);

                for (const filterKey of filtersKeys) {
                    const filterVal = dbFilters[filterKey];

                    // Caller can define some custom functions to generate filter condition (like looking on edges to
                    // filter on libraries linked to an attribute). So, if a custom function is define for this filter,
                    // we use it, otherwise we use the standard filters
                    const filterCondsFunc =
                        typeof customFilterConditions[filterKey] !== 'undefined'
                            ? customFilterConditions[filterKey]
                            : _getFilterCondition;
                    const conds = filterCondsFunc(filterKey, filterVal, strictFilters, nonStrictFields);

                    if (conds?.query) {
                        queryParts.push(aql`FILTER`, conds);
                    }
                }
            }

            if (!!sort) {
                const field = sort.field === 'id' ? '_key' : sort.field;
                queryParts.push(aql`SORT el.${field} ${sort.order}`);
            }

            if (!!pagination) {
                queryParts.push(aql`LIMIT ${pagination.offset || 0}, ${pagination.limit}`);
            }

            queryParts.push(aql`RETURN el`);

            const query = aql.join(queryParts);
            const res = await dbService.execute<IExecuteWithCount | any[]>({query, withTotalCount: withCount, ctx});

            const results = !Array.isArray(res) ? res.results : res;

            return {
                totalCount: withCount ? (res as IExecuteWithCount).totalCount : null,
                list: results.map(ret.cleanup)
            };
        },
        convertValueVersionToDb(version: IValueVersion): IDbValueVersion {
            return Object.keys(version).reduce((allVers, treeName) => {
                const id = version[treeName];

                allVers[treeName] = id;

                return allVers;
            }, {});
        },
        async clearDatabase(): Promise<void> {
            // Drop all collections
            const cols = await dbService.db.listCollections();
            for (const col of cols) {
                const colType =
                    col.type === CollectionType.DOCUMENT_COLLECTION ? collectionTypes.DOCUMENT : collectionTypes.EDGE;
                await dbService.dropCollection(col.name, colType);
            }
        }
    };

    return ret;
}
