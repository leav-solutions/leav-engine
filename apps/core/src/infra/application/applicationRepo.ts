// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {appRootPath} from '@leav/app-root-path';
import {aql} from 'arangojs';
import {readdir} from 'fs/promises';
import path from 'path';
import {IUtils} from 'utils/utils';
import {IConfig} from '_types/config';
import {IList} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {APPS_MODULES_FOLDER, IApplication, IApplicationModule} from '../../_types/application';
import {IGetCoreEntitiesParams} from '../../_types/shared';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';

export interface IApplicationRepo {
    getApplications(params: {params?: IGetCoreEntitiesParams; ctx: IQueryInfos}): Promise<IList<IApplication>>;
    updateApplication(params: {applicationData: IApplication; ctx: IQueryInfos}): Promise<IApplication>;
    createApplication(params: {applicationData: IApplication; ctx: IQueryInfos}): Promise<IApplication>;
    deleteApplication(params: {id: string; ctx: IQueryInfos}): Promise<IApplication>;
    getAvailableModules(params: {ctx: IQueryInfos}): Promise<IApplicationModule[]>;
}

export const APPLICATIONS_COLLECTION_NAME = 'core_applications';

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.db.dbUtils'?: IDbUtils;
    'core.utils'?: IUtils;
    config?: IConfig;
}

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.db.dbUtils': dbUtils = null,
    'core.utils': utils = null,
    config = null
}: IDeps = {}): IApplicationRepo {
    return {
        async getApplications({params, ctx}): Promise<IList<IApplication>> {
            const defaultParams: IGetCoreEntitiesParams = {
                filters: null,
                strictFilters: false,
                withCount: false,
                pagination: null,
                sort: null
            };

            const initializedParams = {...defaultParams, ...params};

            return dbUtils.findCoreEntity<IApplication>({
                ...initializedParams,
                collectionName: APPLICATIONS_COLLECTION_NAME,
                nonStrictFields: ['label', 'id', 'endpoint'],
                ctx
            });
        },
        async updateApplication({applicationData, ctx}): Promise<IApplication> {
            const docToInsert = dbUtils.convertToDoc(applicationData);

            const col = dbService.db.collection(APPLICATIONS_COLLECTION_NAME);
            const res = await dbService.execute({
                query: aql`UPDATE ${docToInsert} IN ${col} OPTIONS { mergeObjects: false } RETURN NEW`,
                ctx
            });

            return dbUtils.cleanup(res.pop());
        },
        async createApplication({applicationData, ctx}): Promise<IApplication> {
            const docToInsert = dbUtils.convertToDoc(applicationData);
            // Insert in applications collection
            const col = dbService.db.collection(APPLICATIONS_COLLECTION_NAME);
            const res = await dbService.execute({
                query: aql`INSERT ${docToInsert} IN ${col} RETURN NEW`,
                ctx
            });

            return dbUtils.cleanup(res.pop());
        },
        async deleteApplication({id, ctx}): Promise<IApplication> {
            // Delete attribute
            const col = dbService.db.collection(APPLICATIONS_COLLECTION_NAME);

            const res = await dbService.execute({
                query: aql`REMOVE ${{_key: id}} IN ${col} RETURN OLD`,
                ctx
            });

            // Return deleted attribute
            return dbUtils.cleanup(res.pop());
        },
        async getAvailableModules(): Promise<IApplicationModule[]> {
            const rootPath = appRootPath();
            const appRootFolder = path.resolve(rootPath, config.applications.rootFolder, APPS_MODULES_FOLDER);

            const appsFolders = await readdir(appRootFolder);

            const components: IApplicationModule[] = await Promise.all(
                appsFolders.map(async appFolder => {
                    const appPath = path.resolve(appRootFolder, appFolder);
                    const appPackageJson = await import(path.resolve(appPath, 'package.json'));

                    return {
                        id: appPackageJson.name,
                        description: appPackageJson.description,
                        version: appPackageJson.version
                    };
                })
            );

            return components;
        }
    };
}
