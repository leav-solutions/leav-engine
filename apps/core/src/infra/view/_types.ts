// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IList} from '_types/list';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IQueryInfos} from '../../_types/queryInfos';

export const VIEWS_COLLECTION_NAME = 'core_views';

export interface IViewRepo {
    createView(view: IView, ctx: IQueryInfos): Promise<IView>;
    updateView(view: IView, ctx: IQueryInfos): Promise<IView>;
    getViews(params: IGetCoreEntitiesParams, ctx: IQueryInfos): Promise<IList<IView>>;
    deleteView(viewId: string, ctx: IQueryInfos): Promise<IView>;
}

export interface IViewRepoDeps {
    'core.infra.db.dbService': IDbService;
    'core.infra.db.dbUtils': IDbUtils;
}
