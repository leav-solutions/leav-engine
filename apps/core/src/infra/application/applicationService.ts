// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {appRootPath} from '@leav/app-root-path';
import {aql} from 'arangojs';
import {exec} from 'child_process';
import {R_OK} from 'constants';
import fs from 'fs/promises';
import {IDbService} from 'infra/db/dbService';
import path from 'path';
import {IUtils} from 'utils/utils';
import winston from 'winston';
import {IConfig} from '_types/config';
import {IQueryInfos} from '_types/queryInfos';
import {
    ApplicationInstallStatuses,
    APPS_INSTANCES_FOLDER,
    APPS_MODULES_FOLDER,
    IApplication,
    IApplicationInstall
} from '../../_types/application';
import {IApplicationRepo} from './applicationRepo';

export interface IApplicationService {
    runInstall(params: {application: IApplication; ctx: IQueryInfos}): Promise<IApplicationInstall>;
    runUninstall(params: {application: IApplication; ctx: IQueryInfos}): Promise<boolean>;
    runInstallAll(): void;
    runUninstallAll(): void;
}

export const APPLICATION_INSTALL_SCRIPT_NAME = 'app_install.sh';
export const APPLICATION_UNINSTALL_SCRIPT_NAME = 'app_uninstall.sh';

interface IDeps {
    'core.infra.application'?: IApplicationRepo;
    'core.utils'?: IUtils;
    'core.utils.logger'?: winston.Winston;
    config?: IConfig;
}

export default function ({
    'core.infra.application': applicationRepo = null,
    'core.utils': utils = null,
    'core.utils.logger': logger = null,
    config
}: IDeps = {}): IApplicationService {
    const _execCommand = (scriptPath: string, env: {}): Promise<{exitCode: number; out: string}> => {
        return new Promise((resolve, reject) => {
            const child = exec(scriptPath, {env: {...process.env, ...env}, cwd: path.dirname(scriptPath)});
            let out = '';

            child.stdout.on('data', data => {
                out += data;
            });
            child.stderr.on('data', data => {
                out += data;
            });

            child.on('error', err => {
                reject(err);
            });

            child.on('close', exitCode => {
                if (exitCode) {
                    reject(out);
                }

                resolve({exitCode, out});
            });
        });
    };

    const _getInstancesFolder = (rootPath: string): string => {
        return path.resolve(rootPath, config.applications.rootFolder, APPS_INSTANCES_FOLDER);
    };

    const _getDestinationFolder = (applicationId: string, rootPath: string): string => {
        return path.resolve(_getInstancesFolder(rootPath), applicationId);
    };

    const _createInstanceFolder = async (instanceFolderPath: string) => {
        try {
            await fs.mkdir(instanceFolderPath);
        } catch (err) {
            if (err.code === 'EEXIST') {
                return;
            }

            logger.error(err);
            throw err;
        }
    };

    const _runInstall = async ({application}): Promise<IApplicationInstall> => {
        const rootPath = appRootPath();
        const appFolder = path.resolve(
            rootPath,
            config.applications.rootFolder,
            APPS_MODULES_FOLDER,
            application.module
        );
        const scriptPath = `${appFolder}/${APPLICATION_INSTALL_SCRIPT_NAME}`;

        try {
            const fd = await fs.open(scriptPath, 'r');
            await fd.close();
        } catch (err) {
            return {
                status: ApplicationInstallStatuses.ERROR,
                lastCallResult: err.message
            };
        }

        // Make sure instances folder exists
        const instanceFolderPath = _getInstancesFolder(rootPath);
        try {
            await fs.stat(instanceFolderPath);
        } catch (err: any) {
            if (err.code === 'ENOENT') {
                await _createInstanceFolder(instanceFolderPath);
            } else {
                logger.error(err);
                throw err;
            }
        }

        // Define env variables
        const leavEnv = {
            LEAV_API_URL: `${config.server.publicUrl}/${config.server.apiEndpoint}`,
            LEAV_AUTH_URL: `${config.server.publicUrl}/auth/authenticate`,
            LEAV_DEFAULT_LANG: config.lang.default,
            LEAV_AVAILABLE_LANG: config.lang.available.join(','),
            LEAV_LOGIN_ENDPOINT: utils.getFullApplicationEndpoint('login'),
            LEAV_APP_ENDPOINT: utils.getFullApplicationEndpoint(application.endpoint),
            LEAV_APPLICATION_ID: application.id,
            LEAV_DEST_FOLDER: _getDestinationFolder(application.id, rootPath)
        };
        const timer = logger.startTimer();
        try {
            logger.info(`START build of ${application.id}`);
            const res = await _execCommand(`${scriptPath}`, leavEnv);
            timer.done(`END build of ${application.id}, `);
            return {status: ApplicationInstallStatuses.SUCCESS, lastCallResult: String(res.out)};
        } catch (err) {
            timer.done(`ERROR build of ${application.id}, `);
            logger.error(String(err));
            return {status: ApplicationInstallStatuses.ERROR, lastCallResult: String(err)};
        }
    };

    const _runUninstall = async ({application}): Promise<boolean> => {
        const rootPath = appRootPath();
        const appFolder = path.resolve(
            rootPath,
            config.applications.rootFolder,
            APPS_MODULES_FOLDER,
            application.module
        );
        const destinationFolder = _getDestinationFolder(application.id, rootPath);

        // Check if an uninstall script exists. If so, run it before removing the folder
        const scriptPath = `${appFolder}/${APPLICATION_UNINSTALL_SCRIPT_NAME}`;

        let doesScriptExist: boolean;
        try {
            await fs.access(scriptPath, R_OK);
            doesScriptExist = true;
        } catch (err) {
            doesScriptExist = false;
        }
        const timer = logger.startTimer();
        if (doesScriptExist) {
            const leavEnv = {
                LEAV_APPLICATION_ID: application.id,
                LEAV_DEST_FOLDER: _getDestinationFolder(application.id, rootPath)
            };
            logger.info(`START uninstall of ${application.id}`);
            await _execCommand(`${scriptPath}`, leavEnv);
        }

        // Remove destination folder
        await fs.rm(destinationFolder, {recursive: true, force: true});
        timer.done(`END uninstall of ${application.id}, `);
        return true;
    };

    return {
        runInstall: _runInstall,
        runUninstall: _runUninstall,
        async runInstallAll() {
            const ctx: IQueryInfos = {
                userId: '1',
                queryId: 'applicationService_runInstallAll'
            };
            const allApps = await applicationRepo.getApplications({ctx});
            logger.info(
                `${allApps.list.length} application(s) to build (${allApps.list
                    .map(application => application.id)
                    .join(', ')})`
            );
            for (const application of allApps.list) {
                const res = await _runInstall({application});
            }
        },
        async runUninstallAll() {
            const ctx: IQueryInfos = {
                userId: '1',
                queryId: 'applicationService_runUninstallAll'
            };
            const allApps = await applicationRepo.getApplications({ctx});
            logger.info(
                `${allApps.list.length} application(s) to uninstall (${allApps.list
                    .map(application => application.id)
                    .join(', ')})`
            );
            for (const application of allApps.list) {
                const res = await _runUninstall({application});
            }
        }
    };
}
