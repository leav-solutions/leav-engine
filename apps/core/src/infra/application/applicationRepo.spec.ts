// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {Stats} from 'fs';
import fs from 'fs/promises';
import {IDbUtils} from 'infra/db/dbUtils';
import path from 'path';
import winston from 'winston';
import {IConfig} from '_types/config';
import {mockApplication} from '../../__tests__/mocks/application';
import {mockCtx} from '../../__tests__/mocks/shared';
import applicationRepo from './applicationRepo';

describe('applicationRepo', () => {
    const docAppData = {
        _key: 'test_application',
        type: 'internal',
        system: true,
        label: {fr: 'Test'},
        endpoint: 'my-application',
        description: 'Super application',
        libraries: ['products', 'categories'],
        color: 'orange',
        module: 'data-studio'
    };
    const applicationData = {
        ...mockApplication,
        id: 'test_application',
        system: true,
        label: {fr: 'Test'},
        endpoint: 'my-application',
        description: {fr: 'Super application'},
        libraries: ['products', 'categories'],
        trees: ['files', 'categories'],
        color: 'orange',
        module: 'data-studio'
    };

    describe('getApplications', () => {
        test('Get all applications', async () => {
            const mockDbServ = {db: null, execute: global.__mockPromise([])};
            const mockDbUtils: Mockify<IDbUtils> = {
                findCoreEntity: global.__mockPromise([applicationData])
            };

            const repo = applicationRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const trees = await repo.getApplications({ctx: mockCtx});

            expect(mockDbUtils.findCoreEntity.mock.calls.length).toBe(1);
            expect(trees).toEqual([applicationData]);
        });
    });

    describe('createApplication', () => {
        test('Create new application', async () => {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([docAppData])
            };

            const mockCleanupRes = applicationData;
            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(mockCleanupRes),
                convertToDoc: jest.fn().mockReturnValue(docAppData)
            };

            const appRepo = applicationRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const createdAttr = await appRepo.createApplication({applicationData, ctx: mockCtx});
            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/^INSERT/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(createdAttr).toMatchObject(applicationData);
        });
    });

    describe('updateApplication', () => {
        test('Update application', async () => {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([docAppData])
            };

            const mockCleanupRes = applicationData;
            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(mockCleanupRes),
                convertToDoc: jest.fn().mockReturnValue(docAppData)
            };

            const appRepo = applicationRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const updatedApp = await appRepo.updateApplication({applicationData, ctx: mockCtx});
            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/^UPDATE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(updatedApp).toMatchObject(applicationData);
        });
    });

    describe('deleteApplication', () => {
        test('Delete application', async () => {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([docAppData])
            };

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(applicationData),
                convertToDoc: jest.fn().mockReturnValue(docAppData)
            };

            const appRepo = applicationRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            await appRepo.deleteApplication({id: applicationData.id, ctx: mockCtx});

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object');
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/^REMOVE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
        });
    });

    describe('getAvailableModules', () => {
        const mockConfig: Mockify<IConfig> = {
            applications: {rootFolder: '/some/path'}
        };

        afterAll(() => {
            jest.resetAllMocks();
        });

        test('Return modules found on directory', async () => {
            const pathSpy = jest.spyOn(path, 'resolve').mockReturnValueOnce('/some/path');
            const fsSpy = jest.spyOn(fs, 'readdir').mockResolvedValueOnce(['data-studio', 'admin'] as any[]);
            const fsSpyStat = jest.spyOn(fs, 'stat').mockResolvedValue({} as Stats);

            jest.mock(
                '/some/path/data-studio/manifest.json',
                () => ({
                    name: 'data-studio',
                    description: 'data studio description',
                    version: '42'
                }),
                {virtual: true}
            );

            jest.mock(
                '/some/path/admin/manifest.json',
                () => ({
                    name: 'admin',
                    description: 'admin description',
                    version: '42'
                }),
                {virtual: true}
            );

            const repo = applicationRepo({config: mockConfig as IConfig});

            const modules = await repo.getAvailableModules({ctx: mockCtx});

            expect(modules).toEqual([
                {id: 'data-studio', description: 'data studio description', version: '42'},
                {id: 'admin', description: 'admin description', version: '42'}
            ]);

            pathSpy.mockRestore();
            fsSpy.mockRestore();
            fsSpyStat.mockRestore();
        });

        test('Ignore invalid folders', async () => {
            const pathSpy = jest.spyOn(path, 'resolve').mockReturnValueOnce('/some/path');
            const fsSpy = jest.spyOn(fs, 'readdir').mockResolvedValueOnce(['data-studio', 'invalid_module'] as any[]);
            const fsSpyStat = jest.spyOn(fs, 'stat').mockImplementation(statPath => {
                if (String(statPath).match(/invalid_module/)) {
                    throw new Error('Invalid module');
                }
                return Promise.resolve({} as Stats);
            });

            jest.mock(
                '/some/path/data-studio/manifest.json',
                () => ({
                    name: 'data-studio',
                    description: 'data studio description',
                    version: '42'
                }),
                {virtual: true}
            );

            const mockLogger: Mockify<winston.Winston> = {
                warn: jest.fn()
            };

            const repo = applicationRepo({
                'core.utils.logger': mockLogger as winston.Winston,
                config: mockConfig as IConfig
            });

            const modules = await repo.getAvailableModules({ctx: mockCtx});

            expect(modules).toEqual([{id: 'data-studio', description: 'data studio description', version: '42'}]);

            pathSpy.mockRestore();
            fsSpy.mockRestore();
            fsSpyStat.mockRestore();
        });
    });
});
