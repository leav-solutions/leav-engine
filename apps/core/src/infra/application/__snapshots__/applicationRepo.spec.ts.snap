// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applicationRepo createApplication Create new application 1`] = `"INSERT @value0 IN @@value1 RETURN NEW"`;

exports[`applicationRepo createApplication Create new application 2`] = `
{
  "@value1": "core_applications",
  "value0": {
    "_key": "test_application",
    "color": "orange",
    "description": "Super application",
    "endpoint": "my-application",
    "label": {
      "fr": "Test",
    },
    "libraries": [
      "products",
      "categories",
    ],
    "module": "explorer",
    "system": true,
    "type": "internal",
  },
}
`;

exports[`applicationRepo deleteApplication Delete application 1`] = `"REMOVE @value0 IN @@value1 RETURN OLD"`;

exports[`applicationRepo deleteApplication Delete application 2`] = `
{
  "@value1": "core_applications",
  "value0": {
    "_key": "test_application",
  },
}
`;

exports[`applicationRepo updateApplication Update application 1`] = `"UPDATE @value0 IN @@value1 OPTIONS { mergeObjects: false } RETURN NEW"`;

exports[`applicationRepo updateApplication Update application 2`] = `
{
  "@value1": "core_applications",
  "value0": {
    "_key": "test_application",
    "color": "orange",
    "description": "Super application",
    "endpoint": "my-application",
    "label": {
      "fr": "Test",
    },
    "libraries": [
      "products",
      "categories",
    ],
    "module": "explorer",
    "system": true,
    "type": "internal",
  },
}
`;
